src/api/schema/auth.rs

```diff
@@ -0,0 +1,98 @@
+use serde::{Serialize, Deserialize};
+
+/// Represents a way that the user may log in.
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+pub struct AuthMethod {
+    pub name: String,
+    /// The fields they can or must input when authenticating.
+    /// In the event of anonymous browsing, this can be empty.
+    pub fields: Vec<Field>,
+}
+
+/// Represents a field type. Used to allow input validation.
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+pub enum FieldType {
+    Integer,
+    String,
+    Url
+}
+
+/// A field in a login method.
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+pub struct Field {
+    pub name: String,
+    pub field_type: FieldType,
+    pub value: Option<String>, // Populated when 
+    pub required: bool,
+    // Sensitive fields have their value hidden
+    pub sensitive: bool
+}
+
+/// The possible values for the AuthResult.
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+pub enum AuthResult {
+    Success,
+    FailRejected,
+    FailConnectionError,
+    Connecting
+}
+
+// The actual auth instruction and response instruction.
+
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+pub struct AuthAccountInstruction {
+    used_authmethod: AuthMethod,
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+pub struct AuthAccountResponse {
+    // TODO: Account ID
+    result: AuthResult,
+    details: String,
+}
+
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    // Serialization + Deserialization tests
+    // For all of the types, these tests serialize and deserialize them to
+    // ensure it behaves as expected
+    // To see the serialized structs as json when you run the tests, run it
+    // as `cargo test -- --nocapture`
+    #[test]
+    fn test_auth_account_instruction_serialization() {
+        let original = AuthAccountInstruction {
+            used_authmethod: AuthMethod {
+                name: "test".to_string(),
+                fields: vec![],
+            },
+
+        };
+        let serialized = serde_json::to_string(&original).unwrap();
+
+        println!("serialized AuthAccountInstruction = {}", serialized);
```

Comment:
Replace println! with Rust's `log` crates (using debug! or trace! more likely)

Reply:
Looks like it's not supported in test cases. https://stackoverflow.com/questions/67087597/is-it-possible-to-use-rusts-log-info-for-tests

Reply:
Fixed as discussed in external discussion.